cheri-c-programming.tex

```diff
@@ -1213,15 +1214,28 @@ \subsection{Restricting capability properties}
   new capability.
 
 \item[\ccode{void *cheri\_csetbounds(void *c, size\_t x)}] Narrow the bounds of capability
-  \cvar{c} so that the lower bound is the current pointer value (which may
+  \cvar{c} so that the lower bound is the current address (which may
   have been increased relative to \cvar{c}'s original lower bound), and its
   upper bound is suitable for a length of \cvar{x}.
-%  This macro wraps the compiler built-in
+
+  Note that the effective bounds of the returned capability may be
+  wider than the range [\ccode{cheri\_getaddress(c)},
+  \ccode{cheri\_getaddress(c) + x}) due to capability compression (see
+  \Cref{sec:bounds_alignment}).
+% This macro wraps the compiler built-in
 %  \cfunc{\_\_builtin\_cheri\_bounds\_set}
 
 \item[\ccode{void *cheri\_setaddress(void *c, vaddr\_t a)}] Return a new capability with the same permissions and bounds as \cvar{c} with the address set to \cvar{a}.
 This builtin can be useful to re-derive a valid pointer from an address.
-%  This macro wraps the compiler built-in
+
+\ccode{cheri\_setaddress} is able to set an address \cvar{a} that it
+is outside of the current bounds of \cvar{c}.  Dereferencing the
+resulting capability without bringing the address back in bounds will
+fail.  However, if the address \cvar{a} is not representable in the
```

Comment:
In the strictest sense, this isn’t quite true, as you can do something like:

`char *bp = b[10];`
`bp+=15;`
`bp[-7] = ‘\0’l`

I.e., although bp itself is out-of-bounds, the immediate or another register operand supported by the addressing mode of the instruction may allow the access to be brought within bounds. I’m not sure if we want to capture that level of subtlety here, however, but the key idea is that accesses made via a capability must be within its bounds, even if the address of the capability itself is not.

Reply:
"... will fail (but dereference instructions typically take additional offsetting integer register and/or immediate parameters, which may bring the address back in bounds for that particular dereference without altering the capability itself)" ?

Reply:
The model I have in my mind is that `bp[-7]` creates a temporary capability whose address  is in bounds when it is dereferenced.  Even without CHERI I would think of `bp` and `bp - 7` as different pointers, and it's only the latter that has to be in bounds in this case.  However, I'll try to find a balance between being super wordy and taking offsets into account.

Reply:
What I've done for now is just add a cross-reference to the out-of-bounds pointer section.  I will further refine that a section in a followup change.